hawkular-alerts-engine/src/main/resources/org/hawkular/alerts/engine/rules/ConditionMatch.drl

```diff
@@ -314,44 +317,81 @@ rule ProvideDefaultDampening
         $t : Trigger( $tenantId : tenantId, $tid : id, $tmode : mode )
         not Dampening( tenantId == $tenantId, triggerId == $tid, triggerMode == $tmode )
     then
-        if (log != null && log.isDebugEnabled()) {
-            log.debug("Adding default " + $tmode + " dampening for trigger! " + $t.getId());
-        }
+        log.debugf("Adding default %s dampening for trigger! %s", $tmode, $t.getId());
+
         Dampening d = Dampening.forStrict( $tenantId, $tid, $tmode, 1 );
         insert( d );
 end
 
-rule CollectConditionEvals
+// If we have multiple ConditionEvals for the same [multi-condition] Trigger, group them together in
+// a TriggerConditionEvalSet that can then be used to update dampening.
+//
+// Note that these two rules will fire before the dampening rules due to complexity and recency. 
+
+rule ProvideTriggerConditionEvalSet
     when
-        $t  : Trigger( $tenantId : tenantId, $tid : id, $tmode : mode )
-        $ces : TriggerConditionEval( size > 0 )
-               from accumulate(
-                $ce: ConditionEval ( tenantId == $tenantId,  triggerId == $tid ),
-                init( TriggerConditionEval tce = new TriggerConditionEval($tenantId, $tid, $tmode); ),
-                action ( tce.addConditionEval($ce); ),
-                result( tce )
-               )
+        $t   : Trigger( $tenantId : tenantId, $tid : id, $tmode : mode )
+        not TriggerConditionEvalSet( tenantId == $tenantId, triggerId == $tid )
+        $ce1 : ConditionEval ( tenantId == $tenantId,  triggerId == $tid )
+        $ce2 : ConditionEval ( tenantId == $tenantId,  triggerId == $tid, this != $ce1 )
     then
-        $ces.getConditionEvals().stream().forEach(ce -> retract(ce));
-        insert( $ces );
+        retract( $ce1 )
+        retract( $ce2 )
+
+        TriggerConditionEvalSet tces = new TriggerConditionEvalSet( $tenantId, $tid, $ce1, $ce2 );
+        insert( tces );
+
+        log.debugf("ProvideTriggerConditionEvalSet %s", tces);
 end
 
-rule DampenTrigger
+rule AddToTriggerConditionEvalSet

```

Comment:
I am not sure about this one.
A potential problem could happen:
- In a situation where we have a large # of conditions, imagine a (far cornercase but possible) trigger with 5 conditions.
- TriggerConditionEvalSet is inserted in "ProvideTriggerConditionEvalSet" but populated in "addToTriggerConditionEvalSet", and the object in the working memory is not updated.
- At the point is inserted "DampenTriggerMultiEval" is chosen to fire, but there is a potential race where you can fire "DampenTriggerMultiEval" before that all conditions eval have been populated.

The previous "CollectConditionEvals" populate the TriggerConditionEval atomically, so that race situation is addressed.

Certainly, having an improvement for the single ConditionEval situation is good, as we know that multiconditions are less used.

Perhaps these rules can be arranged mixing both approaches, having a case for single ConditionEval, but in the case of multi-conditions, using a rule to accumulate in a single step, not using a strategy of separate rules, perhaps that could work with salience attributes, but I'd prefer if we could avoid them and just using an accumulate sentence for this case.